nurse_scheduling.py

```diff
@@ -0,0 +1,171 @@
+# constraint - value provided by Ikeda, Nakamura, Humble
+a = 3.5
+n_nurses = 3
+n_days = 11
+size = n_days * n_nurses
+
+# Hard shift constraint: at least one nurse working every day
+# Lagrange parameter, for hard nurse constraint, on workforce and effort
+# Workforce function W(d) - set to a constant 'workforce' for now
+# Effort function E(n) - set to a constant 'effort' for now
+lagrange_parameter = 1.3
+workforce = 1
+effort = 1
+
+# Soft nurse constraint: all nurses should have approximately even work
+#                        schedules
+# Lagrange parameter, for shift constraints, on work days
+# preference function 'preference' - set to a constant for now
+# Minimum duty days 'min_duty_days' - the number of work days that each
+# nurse wants
+# to be scheduled. At present, each will do the minimum on average.
+# The parameter gamma's value suggested by Ikeda, Nakamura, Humble
+gamma = 0.3
+preference = 1
+min_duty_days = int(n_days/n_nurses)
+
+
+# Find index into 1D list for (nurse_index, day_index)
+def get_index(nurse_index, day_index):
+    return nurse_index * n_days + day_index
+
+
+# Inverse of get_index - given an index in a 1D list, return the nurse_index
+# and day_index
+def get_nurse_and_day(index):
+    nurse_index, day_index = divmod(index, n_days)
+    return nurse_index, day_index
+
+
+# Hard nurse constraint: no nurse works two consecutive days
+# It does not have Lagrange parameter - instead, J matrix
+# symmetric, real-valued interaction matrix J, whereas all terms are
+# a or zero.
+# composite indices i(n, d) and j(n, d) as functions of n and d
+# In other words, the indices are not n and d. They are kind of
+# (n+d)**2, and we need to take that into account. This seems to be
+# a clean way to do it.
+# J_i(n,d)j(n,d+1) = a and 0 otherwise.
+J = defaultdict(int)
+for nurse in range(n_nurses):
+    for day in range(n_days - 1):
+        nurse_day_1 = get_index(nurse, day)
+        nurse_day_2 = get_index(nurse, day+1)
+        J[nurse_day_1, nurse_day_2] = a
+
+# Q matrix assign the cost term, the J matrix
+Q = deepcopy(J)
+
+# Hard shift constraint: at least one nurse working every day
+# The sum is over each day.
+# This constraint tries to make (effort * sum(q_i)) equal to workforce,
+# which is set to a constant in this implementation, so that one nurse
+# is working each day.
+# lagrange_parameter * ((sum(effort * q_i(n,d)) - workforce) ** 2)
+for nurse_day_1 in range(size):
+    _, date_index = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the workforce * workforce
+    Q[nurse_day_1, nurse_day_1] += lagrange_parameter * ((effort * effort) - (2 * workforce * effort))
+    for nurse_day_2 in range(size):
+        _, day_index_2 = get_nurse_and_day(nurse_day_2)
+        # Include only the same day, across nurses
+        if (date_index == day_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += 2 * lagrange_parameter * effort * effort
```

Comment:
I don't understand why when the nurses are different, the weight drops the `2 * workforce * effort` term. I also feel like the `Q[nurse_day_1, nurse_day_2] += 2 * ..` is double counting. 

*What I think is happening*
- Solver is trying to minimize the constraint. This occurs, when `sum(effort * q_i(n,d))` is approximately the same as `workforce`
- your `q_i(n,d)` represents the binary value that the solver will find. Hence, the constraint you wrote can be written to
```
lagrange * (sum(effort * q_i) - workforce)**2  # where the sum is the sum of values over a day

# we open up the square and collect like-terms
# note: since 'effort' is a scalar, we can pull it in front of the summation
lagrange * (effort**2 - 2*effort*workforce) * sum(q_i) + workforce**2

# since we are only interested in the weight multiplying `q_i` and not
# the offset, workforce**2, we can drop it.

# therefore, the weight (i.e. bias) we apply to `q_i` is
lagrange * (effort**2 - 2*effort*workforce)
```

Hence, i don't understand why the term `2 * workforce * effort` was dropped.

Reply:
Hello Melody, I have possibly misled you, about the nature of the sum, in this term. I went back to the INH paper, and it looks a little different than I led you to believe. Here is what INH has:

    lagrange * sum_d [ E sum_n q_i(n,d) - W] ** 2

This isn't the same as I might have insinuated,

    lagrange * [ E sum_n sum_d q_i(n, d) - W ] ** 2

This is what limits the sum. I have a photo of this, and I also worked out the algebra in my notebook, which I can send you a photo of, if you'd like.

Reply:
Further explanation on how the constraint matches the code would be greatly appreciated. As well, could you please include this explanation in your PR as this code example is open to the wider dev community, hence they may have the same questions as I did.

Reply:
@m3ller, I am adding more material to the main program. Please tell me more about what you're thinking, for explanations in the PR. Should the PR contain a writeup about the details of the calculations? If so, should that same writeup be available in the files as well?

Reply:
You're welcome. It is way cleaner and getting more Pythonic too.